make manipulators use world coordinates

kevinyamauchi opened the pull request:
As discussed in #163 , switching the manipulators to world coordinates may be beneficial for keeping the size correct when the layers have scale.

The current approach:
- add the manipulator visual to the root of the scene https://github.com/napari-threedee/napari-threedee/pull/168/commits/f09616f9636b4a343929fa76b559cdb2d28f08e0
- make the base manipulator use world coordinates: https://github.com/napari-threedee/napari-threedee/pull/168/commits/03b791724a4c0345a50314bec9d3b0863d75619f
- update the point manipulator to use world coordinates: https://github.com/napari-threedee/napari-threedee/pull/168/commits/3d54cadfe546f7bf3b95e34048f0eb7207cfc5b3
- update the render plane manipulator to use world coordinates. needed to add util functions for transforming the normal vector: https://github.com/napari-threedee/napari-threedee/pull/168#issuecomment-2049333677
- make layer manipulator use world coordinates: https://github.com/napari-threedee/napari-threedee/pull/168/commits/20b7aaca9c0cdcce1bb4841630b5e0d39c716768 https://github.com/napari-threedee/napari-threedee/pull/168/commits/835ccff9e61dcbcbed524e4f8e4be39c87fd2df5

kevinyamauchi commented:
Here's a test script that seems to be working. The point manipulators work be first selecting the point and then moving it using the manipulator. Note that it is working for both the transformed and untransformed layer. Also, the manipulator is the same side. (hint: you have to change layer selection to select on different layers - I always forget).

```python
import napari
import numpy as np


points_data = np.array(
    [
        [0, 0, 0],
        [0, 200, 0],
        [0, 0, 200]
    ]
)

viewer = napari.Viewer(ndisplay=3)
points_layer = viewer.add_points(points_data, size=5, face_color="blue", name="identity")
points_layer_translated = viewer.add_points(
    points_data,
    translate=(20, 20, 20),
    scale=(1.5, 0.5, 0.5),
    size=5,
    face_color="green",
    name="transformed"
)

viewer.window.add_plugin_dock_widget(
    plugin_name="napari-threedee", widget_name="point manipulator"
)

napari.run()

```

kevinyamauchi commented:
@psobolewskiPhD , I almost have it working with the render plane manipulator. The positioning and translation is working. The rotations are not quite right yet, but I think it's close.

Here's an example script in case you want to play with it:

``python
import napari
from skimage import data


viewer = napari.Viewer(ndisplay=3)

# add a volume
blobs = data.binary_blobs(
    length=64, volume_fraction=0.1, n_dim=3
).astype(float)
volume_layer = viewer.add_image(
    blobs, rendering='mip', name='volume', blending='additive', opacity=0.25, scale=(1, 0.5, 0.5)
)

# add the same volume and render as plane
# plane should be in 'additive' blending mode or depth looks all wrong
plane_parameters = {
    'position': (32, 32, 32),
    'normal': (1, 0, 0),
    'thickness': 10,
}

plane_layer = viewer.add_image(
    blobs,
    rendering='average',
    name='plane',
    colormap='bop orange',
    blending='additive',
    opacity=0.5,
    depiction="plane",
    plane=plane_parameters,
    scale=(1, 0.5, 0.5)
)

viewer.window.add_plugin_dock_widget(
    plugin_name="napari-threedee", widget_name="render plane manipulator"
)

napari.run()

```

psobolewskiPhD commented:
Yeah the rotations are a bit off. It look like the manipulator isn't placed along the plane normal so then there is a disconnect between the manipulator rotation and the plane rotation.
Also I noticed that manipulators are no longer hidden when the layer is hidden nor--worse yet-- removed when the layer is removed. This must be cause they arn't part of the layer anymore.
I made a PR to your branch with some callbacks in base_manipulator and then event connections in the manipulator classes: https://github.com/kevinyamauchi/napari-threedee/pull/1
A bit naive perhaps, but it works well on my end.

psobolewskiPhD commented on src/napari_threedee/manipulators/render_plane_manipulator.py:
```diff
@@ -30,15 +31,20 @@ def _update_transform(self):
+        origin_world = self.layer.data_to_world(self.layer.plane.position)
+        self.origin = np.array(origin_world)
+        plane_normal_data = self.layer.plane.normal
+        plane_normal_world = data_to_world_ray(vector=plane_normal_data, layer=self.layer)
```
visually it looks like the manipulator isn't normal to the plane, so maybe the issue lies here?
Isn't the plane.normal relative to the plane.position? While in data_to_world_ray zeroes are used?
To be fair I'd expect a bigger discrepancy though.

kevinyamauchi commented:
Thanks for the PR, @psobolewskiPhD ! I will take a look today.

I realized I was making a mistake with computing the normal vectors. I forgot that we need to treat them a bit special when there is non-uniform scaling: https://www.scratchapixel.com/lessons/mathematics-physics-for-computer-graphics/geometry/transforming-normals.html

kevinyamauchi commented:
@psobolewskiPhD , I think I've fixed the render plane normal issue. Indeed it was because I was transforming the normals incorrectly. Would you mind giving it a go?

There are the commits where I made the change:
- https://github.com/napari-threedee/napari-threedee/pull/168/commits/0a3174bbd9d774e1f27f9d4f05c5c3c4c5490a90
- https://github.com/napari-threedee/napari-threedee/pull/168/commits/f795c6b5ebb59e390a5ecda2d5525c37df022c56

kevinyamauchi commented:
@m-albert, I'm not sure if you're still using the layer manipulators, but just in case, we are changing them such that they are defined in world coordinates because it makes it easier to work with layers with different transformations. It _shouldn't_ change the behavior, but I wanted to flag this in case you notice something weird!

m-albert commented:
Thank you for the ping @kevinyamauchi!

I am using the layer manipulator in the sense that it's super useful for manually providing initial transforms to layers that are then further processed by `napari-stitcher`. For this it'll be great to have the manipulators work in world coordinates and taking affine transforms into account.

Will let you know in case I notice anything weird, thank you!

kevinyamauchi commented:
Perfect, thanks @m-albert !

psobolewskiPhD commented:
> I think we want to also account for the deletion of layers. I don't think we have a layer deleted event on the layers, but we could listen to the layer list events. For example, there's a removed event that we could potentially attach a callback to:

You mean move the callback and envent I added for layerlist.removed to NapariManipulatorBackend rather than having it in base_manipulator? I guess based on your post here:
https://github.com/napari-threedee/napari-threedee/issues/163#issuecomment-2046695158
and explanation of the Backend, it would be better there?

kevinyamauchi commented:
> > I think we want to also account for the deletion of layers. I don't think we have a layer deleted event on the layers, but we could listen to the layer list events. For example, there's a removed event that we could potentially attach a callback to:
> 
> You mean move the callback and envent I added for layerlist.removed to NapariManipulatorBackend rather than having it in base_manipulator? I guess based on your post here: [#163 (comment)](https://github.com/napari-threedee/napari-threedee/issues/163#issuecomment-2046695158) and explanation of the Backend, it would be better there?

Ah sorry, I missed those additions. I think that is a behavior we would want for all manipulators, so I think it is best in a class that all manipulators have so that we don't have to remember to add it to each one. I was thinking the `NapariManipulatorBackend` might be the most natural, but I'm open to suggestions!

psobolewskiPhD commented:
So this is probably just my ignorance, but `base_manipulator` calls NapariManipulatorBackend, so I can't see how to get the manipulator visibility, enable state, etc. from within NapariManipulatorBackend
I think the event connections could be added to `base_manipulator` init, which would propagate to the all manipulators...
Does that make sense?

psobolewskiPhD commented:
I tried that, but looks like layer is None at the base_manipulator level. I can connect the viewer.layerlist event though.
Not sure if it's worth the refactor for just 1 out of 2 events....

kevinyamauchi commented:
Thanks for checking @psobolewskiPhD. Fair enough, let's leave it as is.

In that case, I will go ahead and merge!
